libyul/optimiser/SimplificationRules.cpp

```diff
@@ -36,22 +36,44 @@ using namespace solidity::evmasm;
 using namespace solidity::langutil;
 using namespace solidity::yul;
 
-SimplificationRule<yul::Pattern> const* SimplificationRules::findFirstMatch(
+SimplificationRules::Rule const* SimplificationRules::findFirstMatch(
 	Expression const& _expr,
 	Dialect const& _dialect,
 	map<YulString, AssignedValue> const& _ssaValues
 )
 {
+	SimplificationRules::Rule const* result = nullptr;
 	auto instruction = instructionAndArguments(_dialect, _expr);
 	if (!instruction)
 		return nullptr;
 
 	static SimplificationRules rules;
 	assertThrow(rules.isInitialized(), OptimizerException, "Rule list not properly initialized.");
 
-	for (auto const& rule: rules.m_rules[uint8_t(instruction->first)])
+	result = rules.findFirstMatch(_expr, _dialect, _ssaValues, rules.m_rules[uint8_t(instruction->first)]);
+
+	if (result == nullptr)
+		if (EVMDialect const* evmDialect = dynamic_cast<EVMDialect const*>(&_dialect))
+		{
+			EVMVersion version = evmDialect->evmVersion();
+			rules.addEvmRules(version);
+
+			result = rules.findFirstMatch(_expr, _dialect, _ssaValues, rules.m_evm_rules[version][uint8_t(instruction->first)]);
```

Comment:
Why do you need a new member variable for the specialized rules?

Reply:
I didn't wanted to put the "same" `for` loop twice. But probably it's ok to have these loops.

Reply:
I meant to ask why don't you make all rules specialized on an evm version?